car-bridge/src/main.rs

```diff
@@ -29,58 +41,193 @@ async fn main() -> Result<(), Box<dyn Error>> {
+            IntentEnum::Subscribe(_) => todo!(),
+            _ => chariott
+                .fulfill(fulfill_request.namespace, intent_enum)
+                .await
+                .map(|r| r.into_inner()),
+        }?;
+
+        let mut payload = vec![];
+        response.encode(&mut payload)?;
+
+        Ok(Response {
+            payload,
+            content_type: "application/x-proto+chariott.common.v1.FulfillResponse",
+            is_error: false,
+        })
+    }
+    .await;
+
+    let response = match response {
+        Ok(message) => message,
+        Err(error) => {
+            debug!("Error when handling message: '{error:?}'.");
+
+            let message = ValueMessage { value: Some(ValueEnum::String(format!("{error:?}"))) };
+
+            let mut payload = vec![];
+            if let Err(err) = message.encode(&mut payload) {
+                debug!("Failed to encode error response: '{err}'.");
+            }
+
+            Response {
+                payload,
+                content_type: "application/x-proto+chariott.common.v1.Value",
+                is_error: true,
+            }
+        }
+    };
+
+    let mut properties = Properties::new();
+
+    properties.push_string(PropertyCode::ContentType, response.content_type).unwrap();
+
+    properties
+        .push_string_pair(
+            PropertyCode::UserProperty,
+            "error",
+            if response.is_error { "1" } else { "0" },
+        )
+        .unwrap();
+
+    if let Err(err) = properties
+        .push_binary(PropertyCode::CorrelationData, correlation_information.correlation_data)
+    {
+        debug!("Could not set correlation data in properties: '{err}'.");
+        return;
+    }
+
+    if let Err(err) = response_sender
+        .send((
+            correlation_information.response_topic,
+            MessageBuilder::new().payload(response.payload).properties(properties).qos(QOS_2),
+        ))
+        .await
+    {
+        debug!("Failed to send message: '{err:?}'.");
+    }
+}
+
+struct Response {
+    payload: Vec<u8>,
+    content_type: &'static str,
+    is_error: bool,
+}
+
+struct CorrelationInformation {
+    correlation_data: Vec<u8>,
+    response_topic: String,
+}
+
+trait MqttExt {
+    fn get_correlation_information(&self) -> Result<CorrelationInformation, Error>;
+}
+
+impl MqttExt for MqttMessage {
+    fn get_correlation_information(&self) -> Result<CorrelationInformation, Error> {
+        let correlation_data = self
+            .properties()
+            .get_binary(PropertyCode::CorrelationData)
+            .ok_or_else(|| Error::new("No correlation data found on message."))?;
+
+        let response_topic = self
+            .properties()
+            .get_string(PropertyCode::ResponseTopic)
+            .ok_or_else(|| Error::new("No response topic found on message."))?;
+
+        Ok(CorrelationInformation { correlation_data, response_topic })
+    }
+}
```

Comment:
nit: Since the extension only needs properties, it might be more general to just require that and default the rest of the implementation into the trait:

```suggestion
trait MqttExt {
    fn properties(&self) -> &Properties;

    fn get_correlation_information(&self) -> Result<CorrelationInformation, Error> {
        let correlation_data = self
            .properties()
            .get_binary(PropertyCode::CorrelationData)
            .ok_or_else(|| Error::new("No correlation data found on message."))?;

        let response_topic = self
            .properties()
            .get_string(PropertyCode::ResponseTopic)
            .ok_or_else(|| Error::new("No response topic found on message."))?;

        Ok(CorrelationInformation { correlation_data, response_topic })
    }
}

impl MqttExt for MqttMessage {
    fn properties(&self) -> &Properties {
        self.properties()
    }
}
```